hw/ip/otbn/dv/otbnsim/sim/insn.py

```diff
@@ -1272,6 +1280,483 @@ def execute(self, state: OTBNState) -> None:
+        # Cycle 0:  Reg(Tmp) = 0, Reg(C) = 0
+        # Cycle 1:  Reg(Tmp) = [a*b]_d,     Reg(C) = a*b
+        # Cycle 2:  Reg(Tmp) = [Tmp*R]_d,   Reg(C) = a*b
+        # Cycle 3:  Output   = c + (Tmp)*q mod q = [c + (Tmp)*q]^d (-q)
+        #
+        # The required constants q and R are expected to be in the MOD WSR at following locations:
+        # For 16b: q @ [15:0], R @ [47:32]
+        # For 32b: q @ [31:0], R @ [63:32]
+        # 64b and 128b operations are not suppored in RTL. However, for simulator purposes we
+        # implement the functionality already. The constants are expected at:
+        # For  64b: q @ [63:0], R @ [127:64]
+        # For 128b: q @ [127:0], R @ [255:128]
+        #
+        # To simplify the simulator implementation we first compute the result and then emulate
+        # the register update sequence. We do model only the ACC updates. We do not model the MAC's
+        # internal registers tmp and c.
+        result = 0
+
+        # extract the Montgomery constants q and R
+        mod_raw = state.wsrs.MOD.read_unsigned()
+        mod_offset = 32 if (size == 16 or size == 32) else size
+        mod_q = mod_raw & ((1 << size) - 1)
+        mod_R = (mod_raw >> mod_offset) & ((1 << size) - 1)
+
+        for elem in range(256 // size - 1, -1, -1):
+            elem_a = extract_sub_word(vec_a, size, elem)  # in Montgomery space
+            elem_b = extract_sub_word(vec_b, size, elem)  # in Montgomery space
+
+            # Preconditions for Montgomery reduction
+            assert (mod_q > 0) and (mod_q < 2**size)
+
+            # Montgomery computation
+            elem_c = montgomery_mul(elem_a, elem_b, mod_q, mod_R, size)
+
+            elem_c = elem_c & ((1 << size) - 1)
+            result = (result << size) | elem_c
+
+        result = result & ((1 << 256) - 1)
+
+        # Emulate the register usage. We do not model registers TMP and C.
+        # One 64b chunk computation requires 3 cycles. We perform the following register accesses
+        # Cycle    1: TMP and C are written
+        # Cycle    2: TMP is written, ACC and C unused
+        # Cycle    3: ACC is read and written, TMP and C are read
+        #
+        # We now repeat these 3 cycles for all four 64b chunks (quarter words).
+        # For the first QWord we don't read the ACC as it is cleared to all zeros.
+        # In cycle 12 we also write the result to the destination WDR
+        qword_mask = (1 << 64) - 1
+        # Cycle 1-3
+        yield None
+        yield None
+        acc = 0
+        acc |= result & qword_mask
+        acc = state.wsrs.ACC.write_unsigned(acc)
+        # Cycle 4-12
+        for qword in range(1, 4):
+            yield None
+            yield None
+            yield None
+            acc = state.wsrs.ACC.read_unsigned()
+            current_qword_mask = (((1 << 256) - 1) & (qword_mask << (qword * 64)))
+            acc = acc & ~current_qword_mask  # delete the to be replaced qw
```

Comment:
Not sure if I understand the comment.

Reply:
Set all bits where the current 64b chunk (QWord) will be placed to zero such that we can use a bitwise OR in the next line to update these bits.
```suggestion
            acc = acc & ~current_qword_mask  # set all bits to zero at the location of the to be replaced qw
```